controlplane/kubeadm/internal/workload_cluster_etcd.go

```diff
@@ -130,6 +137,39 @@ func (w *Workload) EtcdIsHealthy(ctx context.Context) (HealthCheckResult, error)
 		return response, errors.Errorf("there are %d healthy etcd pods, but %d etcd members", expectedMembers, len(knownMemberIDSet))
 	}
 
+	// Check etcd cluster alarms
+	etcdClient, err := w.etcdClientGenerator.forNodes(ctx, controlPlaneNodes.Items)
+	if err != nil {
+		return response, err
+	}
+
+	defer etcdClient.Close()
+	alarmList, err := etcdClient.Alarms(ctx)
+	if len(alarmList) > 0 || err != nil {
+		return response, errors.Errorf("etcd cluster has %d alarms", len(alarmList))
+	}
+
+	members, err := etcdClient.Members(ctx)
+	if err != nil {
+		return response, err
+	}
+
+	healthyMembers := 0
+	for _, m := range members {
+		if val, ok := response[m.Name]; ok {
+			if val == nil {
+				healthyMembers++
+			}
+		} else {
+			// There are members in etcd cluster that is not part of controlplane nodes.
+			return response, err
+		}
+	}
+	// TODO: During provisioning, this condition may be set false for a short time until all pods are provisioned, can add additional checks here to prevent this.
+	if healthyMembers < (len(members)/2 + 1) {
+		return response, errors.Errorf("etcd lost its quorum: there are %d healthy members, but %d total etcd members", healthyMembers, len(members))
+	}
+
```

Comment:
Should we have a different method that checks only for alarms instead?

Reply:
I don't quite follow, instead of quorum check you mean?